include/psa/crypto_types.h

```diff
@@ -85,10 +85,30 @@ typedef uint32_t psa_algorithm_t;
  */
 
 /** Encoding of key lifetimes.
+ *
+ * The lifetime of a key indicates where it is stored and what system actions
+ * may create and destroy it.
+ *
+ * Keys with the lifetime #PSA_KEY_LIFETIME_VOLATILE are automatically
+ * destroyed when the application terminates or on a power reset.
+ *
+ * Keys with a lifetime other than #PSA_KEY_LIFETIME_VOLATILE are said
+ * to be _persistent_.
+ * Persistent keys are preserved if the application or the system restarts.
+ * Persistent keys have a key identifier of type #psa_key_id_t.
+ * The application can call psa_open_key() to open a persistent key that
+ * it created previously.
  */
 typedef uint32_t psa_key_lifetime_t;
 
 /** Encoding of identifiers of persistent keys.
+ *
+ * - Applications may freely choose key identifiers in the range
+ *   #PSA_KEY_ID_USER_MIN to #PSA_KEY_ID_USER_MAX.
+ * - Implementations may define additional key identifiers in the range
+ *   #PSA_KEY_ID_VENDOR_MIN to #PSA_KEY_ID_VENDOR_MAX.
+ * - Key identifiers outside these ranges are reserved for future use
+ *   in future versions of this specification.
```

Comment:
I cannot recall - did we decide that key id of zero was valid or always invalid? - or if we should have a key id that is always invalid/null.

Or is this now an application decision whether or not to allocate a key with id of zero? - is this worth mentioning at all?

Reply:
I'm in two minds about id=0. On the one hand, allowing it violates the general principle that 0 is a “safe” invalid value. On the other hand, forbidding it means that you can't recognize a valid key id by a simple bit mask, and you can calculate a key id hash by just taking an _n_-bit hash.

I'd rather keep `PSA_KEY_ID_VENDOR_MIN ≤ 1` to not gratuitously break existing code. Is 1..0x3fffffff a sensible range? Should it be 1..0x40000000? Or am I just imagining things and 1..0x3fffffff (or 1..0x7fffffff) is perfectly fine?

Reply:
How many places does an implementation need to verify that a valid key id has been provided? - just during key creation and key opening it seems, so the need to check both ends of the range rather than just a one sided range isn't going to set off the code bloat alarm.

If we implement ARMmbed/psa-crypto#158 to unify handles and key ids then "zero is special" returns anyway to represent the NULL handle.

Reply:
I decided that being consistent with “0 is invalid”, and protecting against the application forgetting to set a value that it really should have set (which is part of the reason for “0 is invalid”), are more important than making the permitted range exactly describable with a power of 2. So I've changed `PSA_KEY_ID_VENDOR_MIN` to 0.